### DISCUSSION TITLE
`array` and `struct` instance in runtime structure

### DISCUSSION THREAD

**Discussion by q82419** [WebAssembly/gc#454] [Q&A] [answered]
Hello, after overviewing the modified spec, still confused in some expected output of the tests.

Take the [`array.new_fixed` in execution](https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-array-mathsf-array-new-fixed-x-n) for example:

```
9. Let ai be the array instance {type deftype, fields fieldval*}.
10. Let a be the length of S.arrays.
11. Append ai to S.arrays.
12. Return the array reference ref.array a.
```

And in the expected return values of [test](https://github.com/WebAssembly/gc/blob/main/test/core/gc/array.wast#L106-L146):

```
(assert_return (invoke "new") (ref.array))
(assert_return (invoke "new") (ref.eq))
(assert_return (invoke "get" (i32.const 0)) (f32.const 1))
```

Here's the questions:

Q1. The expected return value of `(invoke "new")` is only `(ref.array)`, not a specific value. Is this because the `arrayaddr` is defined by embedder?

Q2. The second invocation of `(invoke "new")` is expected to return the `(ref.eq)`. Why doesn't it be `(ref.array)`? Does the second invocation of `(invoke "new")` return the same reference as the first one?

Q3. Continue from Q2. Does it mean that there must be 1 instance for each defined type when no matter how many times the `array.new` and `struct.new` be invoked (such as the concept of `static` in C/C++)? If so, the array length and initial values are overwritten to the same array instance in every invocation of `array.new` instructions, right?

Thanks for your help!

**Comment by rossberg** [comment, marked as answer]
Q1: `(array.ref)` is a "pattern" in the test assertion language, it is matched by any array reference. There is no way to talk about specific array addresses on the assertion level, so that's the most specific construct provided. If more concrete checks are needed, they have to be coded up inside a Wasm module.

Q2: Similarly, `(array.eq)` is a pattern matched by all references that are subtypes of eqref. This test simply checks that arrays are correctly treated as equality types in assertions. It doesn't imply anything about the identity of the reference. (Reference identity cannot be checked with assertion patterns alone, see Q1.)

Q3: No, struct and array instances are the runtime representations of heap-allocated structs and arrays. Consequently, each *.new instruction generates a fresh one.

**Reply by q82419** [reply]
Thanks!
The `(array.eq)` of the second invocation confused me. I didn’t realize that the expected value is only used to match the runtime type.

**Reply by rossberg** [reply]
Oops, of course I meant `ref.array` and `ref.eq` above. :)

**Reply by q82419** [reply]
Nevermind. I understand your description. Thanks!
